classad/_expression.py

```diff
@@ -0,0 +1,299 @@
+        if key in ["error", "false", "is", "isnt", "parent", "true", "undefined"]:
+            raise ValueError(f"{key} is a reserved name")
+        self._data[key] = value
+
+    def __delitem__(self, key: str) -> None:
+        self._data.pop(key, None)
+
+    def __getitem__(self, key: Iterable) -> Expression:
+        if isinstance(key, str):
+            key = [key]
+        expression = self._data
+        for token in key:
+            token = token.casefold()
+            try:
+                expression = expression[token]
+            except KeyError:
+                return Undefined()
+        return expression
+
+    def __len__(self) -> int:
+        return len(self._data)
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self._data)
+
+    def _evaluate(
+        self, key: Iterable = None, my: "ClassAd" = None, target: "ClassAd" = None
+    ) -> Any:
+        """
+        Perform a matchmaking between an expression defined by the named attribute
+        key in the context of the target ClassAd.
+        """
+        if key is None:
+            return self
+        expression = self[key]
+        new_key = key[:-1]
+        return expression._evaluate(key=new_key, my=self, target=target)
+
+    @classmethod
+    def from_grammar(cls, tokens):
+        result = cls()
+        for token in tokens:
+            result[token[0]] = token[1]
+        return result
+
+    def __eq__(self, other):
+        if type(self) == type(other):
+            return self._data == other._data
+        return False
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}>: {self._data}"
+
+
+class NamedExpression(Expression):
+    @classmethod
+    def from_grammar(cls, tokens):
+        if "super" == tokens:
+            raise NotImplementedError("Super is not supported")
+        result = cls()
+        result._expression = tokens
+        return result
+
+
+class FunctionExpression(Expression):
+    def __init__(self, name, args):
+        super().__init__()
+        self._name = name
+        self._expression = args
+
+    def __eq__(self, other):
+        return (
+            type(self) == type(other)
+            and self._expression == other._expression
+            and self._name == other._name
+        )
+
+    def _evaluate(
+        self, key: Iterable = None, my: "ClassAd" = None, target: "ClassAd" = None
+    ) -> Any:
+        expression = []
+        for element in self._expression:
+            expression.append(element._evaluate(key=key, my=my, target=target))
+        return getattr(_functions, self._name)(*expression)
+
+    @classmethod
+    def from_grammar(cls, tokens):
+        return cls(tokens[0], tokens[1])
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}>: {self._name}{self._expression}"
+
+
+class TernaryExpression(Expression):
+    def _evaluate(
+        self, key: Iterable = None, my: "ClassAd" = None, target: "ClassAd" = None
+    ) -> Any:
```

Comment:
Readability may benefit from unpacking to more sensible names.

```python3
predicate, if_true, if_false = self._expression
```

Reply:
Great idea 👍 Done.